<a href="https://colab.research.google.com/github/PosgradoMNA/actividades-de-aprendizaje-A01793654/blob/main/Actividad_Semana6_Visualizacion.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#**Ciencia y analítica de datos**
## Actividad 6
### Visualizacion

#### Profesor: Jobish Vallikavungal Devassia
#### Profesor Tutor: Roberto Antonio Guevara González

#### Bernardo Mijangos Flores
#### Matricula A01793654

In [2]:
#importamos la librerias necesarias

import numpy as np
import pandas as pd

# Punto 1

Descarga los datos enlace a un sitio externo. y carga el dataset en tu libreta

In [5]:
# obtenemos el path para la lectura de los datos
mypath = "https://raw.githubusercontent.com/PosgradoMNA/Actividades_Aprendizaje-/main/default%20of%20credit%20card%20clients.csv"

# generamos un dataframe con los datos obtenidos de mypath
df = pd.read_csv(mypath, sep=",")
df

ID      X1   X2   X3   X4    X5   X6   X7   X8   X9  ...      X15  \
0          1   20000  2.0  2.0  1.0  24.0  2.0  2.0 -1.0 -1.0  ...      0.0   
1          2  120000  2.0  2.0  2.0  26.0 -1.0  2.0  0.0  0.0  ...   3272.0   
2          3   90000  2.0  2.0  2.0  34.0  0.0  0.0  0.0  0.0  ...  14331.0   
3          4   50000  2.0  2.0  1.0  37.0  0.0  0.0  0.0  0.0  ...  28314.0   
4          5   50000  1.0  2.0  1.0  57.0 -1.0  0.0 -1.0  0.0  ...  20940.0   
...      ...     ...  ...  ...  ...   ...  ...  ...  ...  ...  ...      ...   
29995  29996  220000  1.0  3.0  1.0  39.0  0.0  0.0  0.0  0.0  ...  88004.0   
29996  29997  150000  1.0  3.0  2.0  43.0 -1.0 -1.0 -1.0 -1.0  ...   8979.0   
29997  29998   30000  1.0  2.0  2.0  37.0  4.0  3.0  2.0 -1.0  ...  20878.0   
29998  29999   80000  1.0  3.0  1.0  41.0  1.0 -1.0  0.0  0.0  ...  52774.0   
29999  30000   50000  1.0  2.0  1.0  46.0  0.0  0.0  0.0  0.0  ...  36535.0   

           X16      X17      X18      X19      X20     X21      X22     X23  \
0          0.0      0.0      0.0    689.0      0.0     0.0      0.0     0.0   
1       3455.0   3261.0      0.0   1000.0   1000.0  1000.0      0.0  2000.0   
2      14948.0  15549.0   1518.0   1500.0   1000.0  1000.0   1000.0  5000.0   
3      28959.0  29547.0   2000.0   2019.0   1200.0  1100.0   1069.0  1000.0   
4      19146.0  19131.0   2000.0  36681.0  10000.0  9000.0    689.0   679.0   
...        ...      ...      ...      ...      ...     ...      ...     ...   
29995  31237.0  15980.0   8500.0  20000.0   5003.0  3047.0   5000.0  1000.0   
29996   5190.0      0.0   1837.0   3526.0   8998.0   129.0      0.0     0.0   
29997  20582.0  19357.0      0.0      0.0  22000.0  4200.0   2000.0  3100.0   
29998  11855.0  48944.0  85900.0   3409.0   1178.0  1926.0  52964.0  1804.0   
29999  32428.0  15313.0   2078.0   1800.0   1430.0  1000.0   1000.0  1000.0   

         Y  
0      1.0  
1      1.0  
2      0.0  
3      0.0  
4      0.0  
...    ...  
29995  0.0  
29996  0.0  
29997  1.0  
29998  1.0  
29999  1.0  

[30000 rows x 25 columns]

## Información de los datos

Esta investigación apuntó al caso de los pagos predeterminados de los clientes en Taiwán y compara la precisión predictiva de la probabilidad de incumplimiento entre seis métodos de minería de datos. Desde la perspectiva de la gestión de riesgos, el resultado de la precisión predictiva de la probabilidad de incumplimiento estimada será más valioso que el resultado binario de la clasificación: clientes creíbles o no creíbles.

## Información de atributos:

Esta investigación empleó una variable binaria, pago por defecto (Sí = 1, No = 0), como variable de respuesta. Este estudio revisó la literatura y utilizó las siguientes 23 variables como variables explicativas:

* X1: Monto del crédito otorgado (dólar NT): incluye tanto el crédito de consumo individual como su crédito familiar (complementario).
* X2: Género (1 = masculino; 2 = femenino).
* X3: Educación (1 = posgrado; 2 = universidad; 3 = secundaria; 4 = otros).
* X4: Estado civil (1 = casado; 2 = soltero; 3 = otros).
* X5: Edad (año).
* X6 - X11: Historial de pagos pasados. Hicimos un seguimiento de los registros de pagos mensuales pasados ​​(de abril a septiembre de 2005) de la siguiente manera:
  * X6 = el estado de pago en septiembre de 2005;
  * X7 = el estado de pago en agosto de 2005; . . .;
  * X11 = estado de amortización en abril de 2005.

La escala de medición del estado de amortización es: -1 = pagar debidamente; 1 = retraso en el pago de un mes; 2 = retraso en el pago de dos meses; . . .; 8 = retraso en el pago de ocho meses; 9 = retraso en el pago de nueve meses o más.

* X12-X17: Importe del estado de cuenta (dólar NT).
  * X12 = monto del estado de cuenta en septiembre de 2005;
  * X13 = monto del estado de cuenta en agosto de 2005; . . .;
  * X17 = monto del estado de cuenta en abril de 2005.
* X18-X23: Monto del pago anterior (dólar NT).
  * X18 = monto pagado en septiembre de 2005;
  * X19 = monto pagado en agosto de 2005; . . .;
  * X23 = monto pagado en abril de 2005.


Yeh, I. C. y Lien, C. H. (2009).


# Punto 2

Obten la información del DataFrame con los métodos y propiedades: shape, columns, head(), dtypes, info(), isna()

In [6]:
# obtenemos el total de lineas y columnas del dataframe

print(df.shape)

(30000, 25)


In [9]:
# mostramos las etiquetas de las columnas

df.columns

Index(['ID', 'X1', 'X2', 'X3', 'X4', 'X5', 'X6', 'X7', 'X8', 'X9', 'X10',
       'X11', 'X12', 'X13', 'X14', 'X15', 'X16', 'X17', 'X18', 'X19', 'X20',
       'X21', 'X22', 'X23', 'Y'],
      dtype='object')

In [12]:
# mostramos los primeros 7 registros del dataframe

df.head(7)

ID      X1   X2   X3   X4    X5   X6   X7   X8   X9  ...       X15  \
0   1   20000  2.0  2.0  1.0  24.0  2.0  2.0 -1.0 -1.0  ...       0.0   
1   2  120000  2.0  2.0  2.0  26.0 -1.0  2.0  0.0  0.0  ...    3272.0   
2   3   90000  2.0  2.0  2.0  34.0  0.0  0.0  0.0  0.0  ...   14331.0   
3   4   50000  2.0  2.0  1.0  37.0  0.0  0.0  0.0  0.0  ...   28314.0   
4   5   50000  1.0  2.0  1.0  57.0 -1.0  0.0 -1.0  0.0  ...   20940.0   
5   6   50000  1.0  1.0  2.0  37.0  0.0  0.0  0.0  0.0  ...   19394.0   
6   7  500000  1.0  1.0  2.0  29.0  0.0  0.0  0.0  0.0  ...  542653.0   

        X16       X17      X18      X19      X20      X21      X22      X23  \
0       0.0       0.0      0.0    689.0      0.0      0.0      0.0      0.0   
1    3455.0    3261.0      0.0   1000.0   1000.0   1000.0      0.0   2000.0   
2   14948.0   15549.0   1518.0   1500.0   1000.0   1000.0   1000.0   5000.0   
3   28959.0   29547.0   2000.0   2019.0   1200.0   1100.0   1069.0   1000.0   
4   19146.0   19131.0   2000.0  36681.0  10000.0   9000.0    689.0    679.0   
5   19619.0   20024.0   2500.0   1815.0    657.0   1000.0   1000.0    800.0   
6  483003.0  473944.0  55000.0  40000.0  38000.0  20239.0  13750.0  13770.0   

     Y  
0  1.0  
1  1.0  
2  0.0  
3  0.0  
4  0.0  
5  0.0  
6  0.0  

[7 rows x 25 columns]

In [13]:
# mostramos el tipo de dato de cada columna

df.dtypes

ID       int64
X1       int64
X2     float64
X3     float64
X4     float64
X5     float64
X6     float64
X7     float64
X8     float64
X9     float64
X10    float64
X11    float64
X12    float64
X13    float64
X14    float64
X15    float64
X16    float64
X17    float64
X18    float64
X19    float64
X20    float64
X21    float64
X22    float64
X23    float64
Y      float64
dtype: object

In [14]:
# mostramos informacion mas completa del dataframe

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30000 entries, 0 to 29999
Data columns (total 25 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   ID      30000 non-null  int64  
 1   X1      30000 non-null  int64  
 2   X2      29999 non-null  float64
 3   X3      29998 non-null  float64
 4   X4      29998 non-null  float64
 5   X5      29995 non-null  float64
 6   X6      29997 non-null  float64
 7   X7      29995 non-null  float64
 8   X8      29993 non-null  float64
 9   X9      29991 non-null  float64
 10  X10     29984 non-null  float64
 11  X11     29986 non-null  float64
 12  X12     29989 non-null  float64
 13  X13     29989 non-null  float64
 14  X14     29987 non-null  float64
 15  X15     29985 non-null  float64
 16  X16     29983 non-null  float64
 17  X17     29990 non-null  float64
 18  X18     29992 non-null  float64
 19  X19     29991 non-null  float64
 20  X20     29992 non-null  float64
 21  X21     29989 non-null  float64
 22

In [15]:
# obtenemos el total de valores nulos por cada columna

df.isna().sum()

ID      0
X1      0
X2      1
X3      2
X4      2
X5      5
X6      3
X7      5
X8      7
X9      9
X10    16
X11    14
X12    11
X13    11
X14    13
X15    15
X16    17
X17    10
X18     8
X19     9
X20     8
X21    11
X22    11
X23     5
Y       3
dtype: int64

se observa que en la mayoria de las columnas faltan datos sobre todo desde la columna X10 al X22, por lo que realizaremos una copia del dataframe y trabajaremos con dicha copia.

# Punto 3

Limpia los datos eliminando los registros nulos o rellena con la media de la columna


In [17]:
# realizamos una copia para trabajar en ella

dfn = df.copy()
dfn

ID      X1   X2   X3   X4    X5   X6   X7   X8   X9  ...      X15  \
0          1   20000  2.0  2.0  1.0  24.0  2.0  2.0 -1.0 -1.0  ...      0.0   
1          2  120000  2.0  2.0  2.0  26.0 -1.0  2.0  0.0  0.0  ...   3272.0   
2          3   90000  2.0  2.0  2.0  34.0  0.0  0.0  0.0  0.0  ...  14331.0   
3          4   50000  2.0  2.0  1.0  37.0  0.0  0.0  0.0  0.0  ...  28314.0   
4          5   50000  1.0  2.0  1.0  57.0 -1.0  0.0 -1.0  0.0  ...  20940.0   
...      ...     ...  ...  ...  ...   ...  ...  ...  ...  ...  ...      ...   
29995  29996  220000  1.0  3.0  1.0  39.0  0.0  0.0  0.0  0.0  ...  88004.0   
29996  29997  150000  1.0  3.0  2.0  43.0 -1.0 -1.0 -1.0 -1.0  ...   8979.0   
29997  29998   30000  1.0  2.0  2.0  37.0  4.0  3.0  2.0 -1.0  ...  20878.0   
29998  29999   80000  1.0  3.0  1.0  41.0  1.0 -1.0  0.0  0.0  ...  52774.0   
29999  30000   50000  1.0  2.0  1.0  46.0  0.0  0.0  0.0  0.0  ...  36535.0   

           X16      X17      X18      X19      X20     X21      X22     X23  \
0          0.0      0.0      0.0    689.0      0.0     0.0      0.0     0.0   
1       3455.0   3261.0      0.0   1000.0   1000.0  1000.0      0.0  2000.0   
2      14948.0  15549.0   1518.0   1500.0   1000.0  1000.0   1000.0  5000.0   
3      28959.0  29547.0   2000.0   2019.0   1200.0  1100.0   1069.0  1000.0   
4      19146.0  19131.0   2000.0  36681.0  10000.0  9000.0    689.0   679.0   
...        ...      ...      ...      ...      ...     ...      ...     ...   
29995  31237.0  15980.0   8500.0  20000.0   5003.0  3047.0   5000.0  1000.0   
29996   5190.0      0.0   1837.0   3526.0   8998.0   129.0      0.0     0.0   
29997  20582.0  19357.0      0.0      0.0  22000.0  4200.0   2000.0  3100.0   
29998  11855.0  48944.0  85900.0   3409.0   1178.0  1926.0  52964.0  1804.0   
29999  32428.0  15313.0   2078.0   1800.0   1430.0  1000.0   1000.0  1000.0   

         Y  
0      1.0  
1      1.0  
2      0.0  
3      0.0  
4      0.0  
...    ...  
29995  0.0  
29996  0.0  
29997  1.0  
29998  1.0  
29999  1.0  

[30000 rows x 25 columns]

In [19]:
# procedemos a eliminar los nan y revisaremos cuanto registros se pierden de los 30,000 iniciales
dfn.dropna(inplace = True)

# verificamos si aun existen valores nulos
dfn.isna().values.any()

False

In [20]:
# corroboramos con las columnas

dfn.isnull().any()

ID     False
X1     False
X2     False
X3     False
X4     False
X5     False
X6     False
X7     False
X8     False
X9     False
X10    False
X11    False
X12    False
X13    False
X14    False
X15    False
X16    False
X17    False
X18    False
X19    False
X20    False
X21    False
X22    False
X23    False
Y      False
dtype: bool

In [22]:
# mostramos la copia del dataframe depurada

dfn.shape

(29958, 25)

se observa un total de 42 registros eliminados de los 30,000 iniciales lo que representa el .14% de los datos.  Se puede considerar el uso de la media o mediana para reemplazar los valores nulos pero considero que el .14% no es representativo para los datos y se puede trabajar sin ellos

# Punto 4

Calcula la estadística descriptiva con describe() y explica las medidas de tendencia central y dispersión

In [44]:
# obtenemos la estadistica de X1 y X2

dfn[['X1','X5']].describe()

X1            X5
count    29958.000000  29958.000000
mean    167555.900928     35.483443
std     129737.299088      9.214319
min      10000.000000     21.000000
25%      50000.000000     28.000000
50%     140000.000000     34.000000
75%     240000.000000     41.000000
max    1000000.000000     79.000000

Se obtiene la estadistica de las variables X1 y X5 (Monto del crédito otorgado (dólar NT) y Edad) dado que las demas variables X2, X3 y X4 son de clasificación (Género, Educacion, Estado Civil)

X1 (Monto del crédito otorgado) tiene una media de 167555.9 el 50% de los datos se encuentra entre los montos de 50000 y 240000 y su mediana es de 140000 de un total de 29958 registros el valor minimo es de 10000 y el maximo es de 1000000

X2 (Edad) tiene una media de 35.48, el 50% de los datos se encuentra entre los 28 y 41 su mediana es de 34 de un total de 29958 registros, la edad minima es de 21 y la maxima 79




In [47]:
# obtenemos la estadistica de X12 a X17

dfn[['X12','X13','X14','X15','X16','X17']].describe()

X12            X13           X14            X15  \
count   29958.000000   29958.000000  2.995800e+04   29958.000000   
mean    51248.119901   49200.493825  4.703239e+04   43279.335370   
std     73674.949943   71211.232744  6.938524e+04   64364.684347   
min   -165580.000000  -69777.000000 -1.572640e+05 -170000.000000   
25%      3559.250000    2984.000000  2.664750e+03    2327.500000   
50%     22379.000000   21194.500000  2.008550e+04   19037.500000   
75%     67190.000000   64027.750000  6.018300e+04   54551.250000   
max    964511.000000  983931.000000  1.664089e+06  891586.000000   

                 X16            X17  
count   29958.000000   29958.000000  
mean    40328.984578   38889.925763  
std     60826.219326   59582.883301  
min    -81334.000000 -339603.000000  
25%      1762.250000    1256.000000  
50%     18104.500000   17067.500000  
75%     50220.750000   49234.750000  
max    927171.000000  961664.000000

Revisando la estadistica de X12 a X17 (Importe del estado de cuenta (dólar NT) de los meses de abril a septiembre como se describe se puede observar que la media se encuentra entre los 51248 en X12 y 38889 en X17 asi como la mediana se encuentra en 22379 en X12 y 17067 en X17, dando a entende que hubo un incremento de abril a septiembre

In [49]:
# obtenemos la estadistica de X18 a X23

dfn[['X18','X19','X20','X21','X22','X23']].describe()

X18           X19            X20            X21  \
count   29958.000000  2.995800e+04   29958.000000   29958.000000   
mean     5664.614460  5.925715e+03    5228.429969    4829.873556   
std     16568.823518  2.305598e+04   17617.338167   15676.205514   
min         0.000000  0.000000e+00       0.000000       0.000000   
25%      1000.000000  8.352500e+02     390.000000     296.250000   
50%      2100.000000  2.009000e+03    1800.000000    1500.000000   
75%      5007.000000  5.000000e+03    4511.500000    4014.750000   
max    873552.000000  1.684259e+06  896040.000000  621000.000000   

                 X22            X23  
count   29958.000000   29958.000000  
mean     4801.481574    5220.708025  
std     15285.552652   17788.983767  
min         0.000000       0.000000  
25%       253.250000     118.000000  
50%      1500.000000    1500.000000  
75%      4040.000000    4000.000000  
max    426529.000000  528666.000000

Con los datos de X18 a X23 (Monto del pago anterior) se observa una media de 5664 en X18 y 5220 en X23 asi mismo la mediana es de 2100 en X18 y 1500 en X23 lo que demuestra el incremento de pagos en los meses de abril a septiembre asi el valor maximo es de 873552 en X18 y 528666 en X23

# Punto 5

Realiza el conteo de las variables categóricas

In [70]:

print( pd.value_counts(dfn['X2']) )
print(100 * pd.value_counts(dfn['X2'])/dfn['X2'].count())


2.0    18095
1.0    11863
Name: X2, dtype: int64
2.0    60.401228
1.0    39.598772
Name: X2, dtype: float64


In [71]:

print( pd.value_counts(dfn['X3']) )
print(100 * pd.value_counts(dfn['X3'])/dfn['X3'].count())


2.0    14009
1.0    10572
3.0     4909
5.0      280
4.0      123
6.0       51
0.0       14
Name: X3, dtype: int64
2.0    46.762134
1.0    35.289405
3.0    16.386274
5.0     0.934642
4.0     0.410575
6.0     0.170238
0.0     0.046732
Name: X3, dtype: float64


In [72]:

print( pd.value_counts(dfn['X4']) )
print(100 * pd.value_counts(dfn['X4'])/dfn['X4'].count())


2.0    15939
1.0    13643
3.0      322
0.0       54
Name: X4, dtype: int64
2.0    53.204486
1.0    45.540423
3.0     1.074838
0.0     0.180252
Name: X4, dtype: float64


X2: Género (1 = masculino; 2 = femenino).
X3: Educación (1 = posgrado; 2 = universidad; 3 = secundaria; 4 = otros).
X4: Estado civil (1 = casado; 2 = soltero; 3 = otros).

### 6 Escala los datos, si consideras necesario


### 7 Reduce las dimensiones con PCA, si consideras necesario.


1.   Indica la varianza de los datos explicada por cada componente seleccionado. Para actividades de exploración de los datos la varianza > 70%
2.   Indica la importancia de las variables en cada componente

### 8 Elabora los histogramas de los atributos para visualizar su distribución

### 9 Realiza la visualización de los datos usando por lo menos 3 gráficos que consideres adecuados: plot, scatter, jointplot, boxplot, areaplot, pie chart, pairplot, bar chart, etc.

### 10 Interpreta y explica cada uno de los gráficos indicando cuál es la información más relevante que podría ayudar en el proceso de toma de decisiones.